### PR TITLE
BR: Auto truncate log backup in backup schedule

### PR DESCRIPTION
**Pull request by WizardXiao**
### What is changed and how does it work?

1. add `logBackupTemplate` to backup schedule, it can manage log backup, include start, auto truncate, delete
2. add `logBackup` to  backup schedule's status
3. truncate log backup accroding to `maxReservedTime` like snapshot backup

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 
- [ ] No code 

1. create backup schedule with log backup:
BlTLqhn6oY
5. before backup exceed the time limit
4790c6a1-8f92-4caf-992b-7e4aef52c9a7
6. after backup exceed the time limit
c2906676-1292-42b3-9315-3c701ab9ab00
7. log backup was auto truncated
bc33165d-f0d6-4bc3-9503-f77682cf28d8

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 1]
```diff
@@ -124,6 +130,26 @@ func (bm *backupScheduleManager) canPerformNextBackup(bs *v1alpha1.BackupSchedul
 	return controller.RequeueErrorf("backup schedule %s/%s, the last backup %s is still running", ns, bsName, bs.Status.LastBackup)
 }
 
+func (bm *backupScheduleManager) performLogBackupIfNeeded(bs *v1alpha1.BackupSchedule) error {
+	ns := bs.GetNamespace()
+	bsName := bs.GetName()
+
+	// log backup already run
```
Comment (grovecai) [inline pkg/backup/backupschedule/backup_schedule_manager.go:137]: comment is not correct

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 2]
```diff
@@ -288,26 +357,232 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
...
+
+	for _, backup := range backupsList {
+		if backup.Spec.Mode == v1alpha1.BackupModeLog {
+			logBackup = backup
+			continue
+		}
+		ascBackupList = append(ascBackupList, backup)
+	}
+
+	for i := 0; i < len(ascBackupList)-1; i++ {
```
Comment (grovecai) [inline pkg/backup/backupschedule/backup_schedule_manager.go:423]: can you use sort.Slice()?

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 3]
```diff
@@ -288,26 +357,232 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
...
+			return
+		}
+		klog.Infof("backup schedule %s/%s truncate log backup %s success, truncateTSO %d", ns, bsName, logBackup.GetName(), truncateTSO)
+	}
+
+	if len(expiredBackups) == len(backupsList) && len(expiredBackups) > 0 {
 		// All backups have been deleted, so the last backup information in the backupSchedule should be reset
 		bm.resetLastBackup(bs)
 	}
 }
```
Comment (grovecai) [inline pkg/backup/backupschedule/backup_schedule_manager.go:407]: The following functions only have pure math calculation logic, not any service calls, please write unit test for them

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 4]
```diff
@@ -288,26 +357,232 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
...
+		lastestTSO               uint64
+		expiredTSO               uint64
+		expiredBackups           []*v1alpha1.Backup
+		truncateTSO              uint64
+		isTruncateTSOInLogBackup bool
+		err                      error
+	)
+
+	// caculate latest ts
+	lastestTSO, err = calculateLatestTSO(backupsList, logBackup)
```
Comment (grovecai) [inline pkg/backup/backupschedule/backup_schedule_manager.go:444]: Shouldn't use lastestTSO, we should use NOW, if backup and logbackup stop, we can not clean the expired data

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 2 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go:423]: ok

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 3 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go:407]: ok

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 4 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go:450]: ok, after a discuss, i refine caculate logic and add notes.
https://github.com/pingcap/tidb-operator/pull/4904/files#diff-6d32e105cb814eefb99d8eaa18a5c94f249c6c1ca4e7a9b77b33f6ad218dd479R440

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 5]
```diff
@@ -288,26 +357,232 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
...
+// separateSnapshotBackupsAndLogBackup return snapot backups ordry by create time asc and log backup
+func separateSnapshotBackupsAndLogBackup(backupsList []*v1alpha1.Backup) ([]*v1alpha1.Backup, *v1alpha1.Backup) {
+	var (
+		ascBackupList = make([]*v1alpha1.Backup, 0)
+		logBackup     *v1alpha1.Backup
+	)
+
+	for _, backup := range backupsList {
+		if backup.Spec.Mode == v1alpha1.BackupModeLog {
+			logBackup = backup
```
[the remark this answers is not in the document]
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go:417]: yes, we can find it in backup schedule. the log backup started by backup schedule will record as `status.logBackup` and it has label which is the same as snapshot backup.
https://github.com/pingcap/tidb-operator/pull/4904/files#diff-6d32e105cb814eefb99d8eaa18a5c94f249c6c1ca4e7a9b77b33f6ad218dd479R304

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 1 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go]: ok

File: pkg/apis/pingcap/v1alpha1/backup_schedule.go [hunk 6]
```diff
@@ -21,3 +21,7 @@ import (
 func (bs *BackupSchedule) GetBackupCRDName(timestamp time.Time) string {
 	return fmt.Sprintf("%s-%s", bs.GetName(), timestamp.UTC().Format(BackupNameTimeFormat))
 }
+
+func (bs *BackupSchedule) GetLogBackupCRDName() string {
+	return fmt.Sprintf("%s-%s", "log", bs.GetName())
```
Comment (Ehco1996) [inline pkg/apis/pingcap/v1alpha1/backup_schedule.go:26]:
```suggestion
	return fmt.Sprintf("%s-%s", bs.GetLogBackupName(), bs.GetName())
```

File: pkg/apis/util/config/config.go [hunk 7]
```diff
@@ -388,3 +388,11 @@ func GoTimeToTS(t time.Time) uint64 {
 	ts := (t.UnixNano() / int64(time.Millisecond)) << 18
 	return uint64(ts)
 }
+
+func TransToTS(tso uint64) int64 {
```
Comment (Ehco1996) [inline pkg/apis/util/config/config.go:392]:
```suggestion
func TSOToTS(tso uint64) int64 {
```

File: pkg/apis/util/config/config.go [hunk 8]
```diff
@@ -388,3 +388,11 @@ func GoTimeToTS(t time.Time) uint64 {
...
+	return int64((tso / 1000) >> 18)
+}
+
+func TransToTSO(ts int64) uint64 {
```
Comment (Ehco1996) [inline pkg/apis/util/config/config.go:396]:
```suggestion
func TSToTSO(ts int64) uint64 {
```

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 9]
```diff
@@ -249,6 +275,44 @@ func buildBackup(bs *v1alpha1.BackupSchedule, timestamp time.Time) *v1alpha1.Bac
 	return backup
 }
 
+func buildLogBackup(bs *v1alpha1.BackupSchedule, timestamp time.Time) *v1alpha1.Backup {
+	ns := bs.GetNamespace()
+	bsName := bs.GetName()
+
+	logBackupSpec := *bs.Spec.LogBackupTemplate.DeepCopy()
```
Comment (Ehco1996) [inline pkg/backup/backupschedule/backup_schedule_manager.go:282]: Is it possible to encounter a nil error here? when `LogBackupTemplate` = nil

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 10]
```diff
@@ -288,26 +352,260 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
 		return
 	}
 
-	var deleteCount int
-	for _, backup := range backupsList {
-		if backup.CreationTimestamp.Add(reservedTime).After(bm.now()) {
-			continue
+	ascBackups, logBackup := separateSnapshotBackupsAndLogBackup(backupsList)
```
Comment (Ehco1996) [inline pkg/backup/backupschedule/backup_schedule_manager.go:355]: what is `ascBackups `

File: pkg/apis/util/config/config.go (hunk 7 above)
Comment (WizardXiao) [reply pkg/apis/util/config/config.go:392]: ok

File: pkg/apis/util/config/config.go (hunk 8 above)
Comment (WizardXiao) [reply pkg/apis/util/config/config.go:396]: ok

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 9 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go:282]: ok, but `performLogBackupIfNeeded` will check `LogBackupTemplate`, if it's nil, will not build log backup. add a check here is fine too.

File: pkg/apis/pingcap/v1alpha1/backup_schedule.go (hunk 6 above)
Comment (Ehco1996) [reply pkg/apis/pingcap/v1alpha1/backup_schedule.go:26]: got

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 11]
```diff
@@ -525,24 +527,44 @@ func calculateLatestTSO(backupsList []*v1alpha1.Backup, logBackup *v1alpha1.Back
...
 	)
 
 	for ; i < len(backupsList); i++ {
 		currentBackupTSO, err = config.ParseTSString(backupsList[i].Status.CommitTs)
 		if err != nil {
-			return nil, 0, perrors.Annotatef(err, "parse backup ts of backup %s/%s", backupsList[i].Namespace, backupsList[i].Name)
+			return nil, perrors.Annotatef(err, "parse backup ts of backup %s/%s", backupsList[i].Namespace, backupsList[i].Name)
+		}
+
+		if currentBackupTSO > expiredTSO {
```
Comment (WangLe1321) [inline pkg/backup/backupschedule/backup_schedule_manager.go:549]: if `currentBackupTSO == expiredTSO`, should we delete the backup?

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 12]
```diff
@@ -525,24 +527,44 @@ func calculateLatestTSO(backupsList []*v1alpha1.Backup, logBackup *v1alpha1.Back
...
 // calculateExpiredTSO calculate latestTSO - reservedTime
 func calculateExpiredTSO(latestTSO uint64, reservedTime time.Duration) uint64 {
-	latestTS := config.TransToTS(latestTSO)
+	latestTS := config.TSOToTS(latestTSO)
 	expiredTS := time.Unix(latestTS, 0).Add(-1 * reservedTime).Unix()
-	return config.TransToTSO(expiredTS)
+	return config.TSToTSO(expiredTS)
 }
 
 // calDeleteSnapshotBackupsByExpiredTSO according to expired tso calculate delete backups and truncate tso
```
Comment (WangLe1321) [inline pkg/backup/backupschedule/backup_schedule_manager.go:535]: the comment is outdated

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 13]
```diff
@@ -525,24 +527,44 @@ func calculateLatestTSO(backupsList []*v1alpha1.Backup, logBackup *v1alpha1.Back
...
+	}
+
+	if i != 0 {
+		return backupsList[:i-1], nil
+	}
+
+	return nil, nil
+}
+
+// calDeleteSnapshotBackupsByExpiredTSO according to expired tso calculate delete backups and truncate tso
```
Comment (WangLe1321) [inline pkg/backup/backupschedule/backup_schedule_manager.go:561]: the comment is outdated

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 12 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go]: ok

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 13 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go]: ok

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 14]
```diff
@@ -288,26 +356,260 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
...
+
+// calculateExpiredTSO calculate latestTSO - reservedTime
+func calculateExpiredTSO(latestTSO uint64, reservedTime time.Duration) uint64 {
+	latestTS := config.TSOToTS(latestTSO)
+	expiredTS := time.Unix(latestTS, 0).Add(-1 * reservedTime).Unix()
+	return config.TSToTSO(expiredTS)
+}
+
+// calExpiredSnapshotBackupsWithLogBackup according to expired tso calculate expired backups
+func calExpiredSnapshotBackupsWithLogBackup(backupsList []*v1alpha1.Backup, expiredTSO uint64) ([]*v1alpha1.Backup, error) {
```
Comment (grovecai) [inline pkg/backup/backupschedule/backup_schedule_manager.go:536]: wrong naming?

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 15]
```diff
@@ -288,26 +356,260 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
...
+
+	if i != 0 {
+		return backupsList[:i-1], nil
+	}
+
+	return nil, nil
+}
+
+// calLogBackupExpiredTSO according to expired tso calculate truncate tso
+func calLogBackupExpiredTSO(backupsList []*v1alpha1.Backup, logBackup *v1alpha1.Backup, expiredTSO uint64) (uint64, error) {
```
Comment (grovecai) [inline pkg/backup/backupschedule/backup_schedule_manager.go:562]: wrong naming?

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 16]
```diff
@@ -288,26 +356,260 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
...
+// snapshot1----snapshot2--------------snapshot3-----snapshot...----snapshot n-----------> snapshot backups
+// ---------------------------------------------------------------------checkpointTS----> log backup
+// ---t1-----------t2-------expiredTS-----t3------------t...-----------tn------------------> time
+//
+// supposed that expiredTS = checkpointTS - retentionPeriod, and expiredTS is between t2 and t3.
+// we should promise user can restore to any time between expiredTS and checkpointTS,
+// so we should reserve snapshots from snapshot2 to snapshot n.
+// then we can delete snapshot1 and truncate log backup to t2.
+// so the returned value is snapshot1 and t2.
+func caculateExpiredBackupsWithLogBackup(backupsList []*v1alpha1.Backup, logBackup *v1alpha1.Backup, reservedTime time.Duration) ([]*v1alpha1.Backup, uint64, error) {
```
Comment (grovecai) [inline pkg/backup/backupschedule/backup_schedule_manager.go:439]: caculateExpiredBackupsWithLogBackup => calculateExpiredBackupsAndLogBackup

File: pkg/backup/backupschedule/backup_schedule_manager.go [hunk 17]
```diff
@@ -288,26 +356,260 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
...
+	// caculate latest ts
+	lastestTSO, err = calculateLatestTSO(backupsList, logBackup)
+	if err != nil {
+		return nil, 0, perrors.Annotate(err, "calculate latest tso")
+	}
+
+	expiredTSO = calculateExpiredTSO(lastestTSO, reservedTime)
+
+	// calculate expired backups
+	expiredBackups, err = calExpiredSnapshotBackupsWithLogBackup(backupsList, expiredTSO)
```
Comment (grovecai) [inline pkg/backup/backupschedule/backup_schedule_manager.go:457]: calExpiredSnapshotBackupsWithLogBackup => calExpiredBackupsWithLogBackupOn

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 15 above)
Comment (grovecai) [reply pkg/backup/backupschedule/backup_schedule_manager.go:562]: calLogBackupExpiredTSO => calLogBackupTruncateTSO

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 14 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go]: ok, change it to be `calExpiredBackupsWithLogBackupOn` and add some note.

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 15 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go]: ok

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 16 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go]: ok

File: pkg/backup/backupschedule/backup_schedule_manager.go (hunk 17 above)
Comment (WizardXiao) [reply pkg/backup/backupschedule/backup_schedule_manager.go]: ok

**Review by grovecai — APPROVED** [review]
LGTM

**Review by WangLe1321 — APPROVED** [review]

**Pull request merged**